src/report/ReportPoll.ts

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Mjolnir, REPORT_POLL_EVENT_TYPE } from "../Mjolnir";
+import { ReportManager } from './ReportManager';
+import { LogLevel } from "matrix-bot-sdk";
+
+class InvalidStateError extends Error {}
+
+export class ReportPoll {
+    /*
+     * https://matrix-org.github.io/synapse/latest/admin_api/event_reports.html
+     * "from" is an opaque token that is returned from the API to paginate reports
+     */
+    private from = 0;
+    private timeout: ReturnType<typeof setTimeout> | null = null;
+
+    /*
+     * A class to poll synapse's report endpoint, so we can act on new reports
+     *
+     * @param client The Matrix client underpinning the running Mjolnir
+     * @param manager The report manager in to which we feed new reports
+     * @param save An abstract function to persist where we got to in report reading
+     */
+    constructor(
+        private mjolnir: Mjolnir,
+        private manager: ReportManager,
+    ) { }
+
+    private schedulePoll() {
+        this.timeout = setTimeout(
+            this.getAbuseReports.bind(this),
+            60_000 // a minute in milliseconds
+        );
+    }
+
+    private async getAbuseReports(): Promise<any> {
+        let response_: {
+            event_reports: { room_id: string, event_id: string, sender: string, reason: string }[],
+            next_token: number | undefined
+        } | undefined = undefined;
+        try {
+            response_ = await this.mjolnir.client.doRequest(
+                "GET",
+                "/_synapse/admin/v1/event_reports",
+                { from: this.from.toString() }
+            );
+        } catch (ex) {
+            await this.mjolnir.logMessage(LogLevel.ERROR, "getAbuseReports", `failed to poll events: ${ex}`);
+            this.schedulePoll();
```

Comment:
I think that you should `return` here, no?

Otherwise, I believe that you're going to call `this.schedulePoll()` twice, which feels like a DoS.